### ISSUE TITLE
Censorship resistance

### ISSUE THREAD

**Issue by blurHY** [ipfs/ipfs#419]
In China, IPFS is unusable because of the GFW.
What does GFW do:
 - Block all default bootstrap nodes
 - Set up a node and start collecting ips to ban
What can we do to make it censorship-resistant ?

**Comment by blurHY** [comment]
Location: Shanghai
Port not opened

![image](https://user-images.githubusercontent.com/33171396/62272966-edc4a300-b46e-11e9-8702-72eb3c6464aa.png)

I think the traffic is limited by GFW, it used to be 7MB/s.

**Comment by Stebalien** [comment]
The bootstrapping issue can be solved with some domain fronting (maybe). It will take some work but not an insurmountable amount of work.

However, the more general problem is currently _very_ hard to solve due to the peer to peer nature of IPFS. At the end of the day, the best solution to this issue is to make the underlying libp2p protocols indispensable so that they _can't_ be blocked without harming commerce (like, e.g., HTTPS).

An alternative is to use domain fronting + relays for everything but that's not sustainable as it requires routing all traffic through these relays.

**Comment by blurHY** [comment]
I have used v2ray and https to bypass GFW for about one year, because GFW cannot distinguish legal traffic or not.
There's an better example, resillio sync, and its Chinese competitor, (i forgot the name). It has an enterprise version, so GFW cannot block though it has some illegal usage

**Comment by LeafarDev** [comment]
@blurHY  which company ?

**Comment by blurHY** [comment]
http://www.verysync.com/

It's for enterprise file syncing, and then GFW won't ban it. So we can use it to download movie, etc.

**Comment by Ayms** [comment]
As a coincidence I posted yesterday https://github.com/ipfs/ipfs/issues/439

See also https://github.com/Ayms/node-Tor/issues/2#issuecomment-66150611

node-Tor was not blocked by the GFW despite of the fact that the nodes were in the consensus, in fact we don't know exactly why, maybe because it implements some variants of the protocol compared to usual Tor nodes (see the node-Tor repo)

But for sure http://peersm.com/peersm2 is not blocked (you can try it it transports the Tor protocol over WebSockets connected to our server for the demo), or if it is now it is because our server is blacklisted (because recognized as a Tor node), it's easy to bypass setting up an unknown server not registered in the consensus

Maybe we will implement WebRTC, we think that for p2p projects such as ipfs two hops are enough (because the guards concept do not apply)

**Comment by blurHY** [comment]
@Ayms centralization ?

**Comment by Ayms** [comment]
@blurHY what do you mean? That's the contrary... decentralization

**Comment by blurHY** [comment]
Do you mean tor or what ?

That's centralized and there's no other ways than DHT against censorship. DHT can be also blocked by GFW.

**Comment by Ayms** [comment]
Of course the Tor network is (very) centralized and DHT are great (but can be dangerous if implemented à la bittorrent, see, https://github.com/Ayms/torrent-live) I don't mean the Tor network, I mean the Tor protocol that could apply to projects such as IPFS (and DHT...)

Did you try peersm2 (this one is using the Tor network, that's just a demo app)? Does it pass the GFW?

**Comment by Ayms** [comment]
What? The GFW collecting browsers? Sorry, no, the next step is probably to implement WebRTC because here the target node to which the browser connects to via ws can indeed be banned

**Comment by blurHY** [comment]
Don't forget the network connectivity is terrible in China. They would ban anything as long as it won't affect economy

**Comment by Ayms** [comment]
No they can't, except blocking everything (that's what you mean?), like ws, webrtc, or fingerprinting tor over it

Can you try the link :-) ? That's a 5mn test, circuits created=success,. It was bypassing easily the GFW years ago (as well as node-Tor nodes), some folks from China did try it at that time

**Comment by blurHY** [comment]
> 
> 
> No they can't, except blocking everything (that's what you mean?), like ws, webrtc, or fingerprinting tor over it
> 
> Can you try the link :-) ? That's a 5mn test, circuits created=success,. It was bypassing easily the GFW years ago (as well as node-Tor nodes), some folks from China did try it at that time

Uh, I will try it tomorrow.
But why they can't block it ?
They can block all IPs of users, i.e. all p2p connections, with a whitelist for enterprises

**Comment by Ayms** [comment]
The question is more "why they should block it", of course they can block whatever they like, easy, collect whatever p2p ips, etc, block, easy again, block ws, webrtc, easy too, then finally block everything (I don't know what they consider commerce related or not)? Years ago it was not blocked, would be interesting to see what happens now (I don't know how things work in China but change your ip before the test if you can)

**Comment by TheZ4ro** [comment]
the only way to anti censor is to built a in place bootstrap node function, which can:
1, works based on DHT or KAD like protocal, and obfs like STP/KCP.
2, build in GFW bootstrap node, or totally distributed. no center node.
3, if someone want to become gateway, then he need to register a public IP or port. this is for his own risk.

**Comment by blurHY** [comment]
I wonder if normal nodes with public IPs can be bootstrap nodes.
If yes, **keeping discovered nodes after shutdown** must be enough, like i2p.

I met a friend recently, who said the **characteristics of p2p protocols are very obvious**, obfs or whatever can be the solution. 

Howerver the ultimate question is what China would think about IPFS and the decentralized web.

Will they simply block all the p2p by a whitelist, or do nothing. (the background is i will be working a zeronet-like stuff based on multiple protocols including ipfs)
btw, it won't be 'i', but a new identity.
I'm not very sure if commerce is really a important thing to China when p2p is out of their control, as **the polity differs**.
Hopefully, applied blockchains, like **filecoin, may be irreplacable**

Here I can get about 200 to 800 peers (China Mobile, with relay enabled or using ipv6). Is this too less or just normal ?

**Comment by Ayms** [comment]
China will for sure not be a fan of decentralized p2p, another concept that I proposed [here](https://github.com/ipfs/ipfs/issues/439#issuecomment-593116241) is that peers can introduce themselves outside of a DHT/registration system (ie you connect to a known peer that can introduce unknow peers connected to it, this works also for WebRTC where you don't have to ask a central point for peers introduction), but as we can see this is not a priority for now for IPFS

**Comment by blurHY** [comment]
Yeah, p2p makes it too hard to censor.

But I don't think that easy to ban p2p, although most Chinese p2p users are restricted by NAT.
- There're massive ips outside of GFW
- China cares about normal international commerce outside of GFW, so they don't use a whitelist.
- We can break through the GFW by some obfuscating protocols, like v2ray

Now the only problem is just bootstrapping
- Use github, which is a absolutely irreplaceable, to provide nodes.
- Steganography, can be encoded by AI against detection

**Comment by bertrandfalguiere** [comment]
Some other leads on bootstraping: https://github.com/ipfs/go-ipfs/issues/3908#issuecomment-330046410

**Comment by jessicaschilling** [comment]
Note: Discussion on applications of IPFS are happening over in the [IPFS Forums](https://discuss.ipfs.io/c/ecosystem/applications-of-ipfs/) now ... please continue the discussion there! 

This issue is being moved over to the archived repo https://github.com/ipfs/apps/ for reference.

**Comment by blurHY** [comment]
Is this about applications ?

**Comment by jessicaschilling** [comment]
Applications as in "uses", not applications as in "apps" ... but in any case, we're trying to move discussion of this sort over to the forums. See you there 😊
